File: pkg/src/hankel.f
c ID: hankel.f, last updated 2025-05-26, F.Osorio

cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
      SUBROUTINE hankel_mat(x, y, n, c, ldc, info)
      INTEGER          n, ldc, info
      DOUBLE PRECISION c(ldc,*), x(*), y(*)
c
c     contructs the n-by-n Hankel matrix based on the x vector
c
c     parameters:
c     x       (input) DOUBLE PRECISION array, dimension (n)
c             first column of the Hankel matrix
c     y       (input) DOUBLE PRECISION array, dimension (n)
c             last column of the Hankel matrix, x(n) = y(1)
c     n       (input) INTEGER
c             length of vector x. n > 0.
c     c       (input/output) DOUBLE PRECISION array, dimension (ldc, n)
c             a square matrix. on exit if info = 0, is the Hankel matrix.
c     ldc     (input) INTEGER
c             ldc is the leading dimension of the array c. ldc >= max(1,n).
c     info    (output) INTEGER
c             = 0: successful exit
c             < 0: if info = -i, the i-th argument had an illegal value
c
c     .. local scalars ..
      INTEGER          i, j, k
c
c     test the input parameters
c
      info = 0
      if (n .LT. 0) then
        info = -2
      else if (ldc .LT. max(1, n)) then
        info = -4
      end if
      if (info .NE. 0) return
c
c     quick return if possible
c
      if ((n .EQ. 0) .OR. (ldc .EQ. 0)) return
c
c     setup Hankel matrix c
c
      do i = 1, n
        do j = 1, n
          k = i + j - 1
          if (k .GT. n) then 
            c(i,j) = y(i + j - n)
          else 
            c(i,j) = x(i + j - 1)
          end if
        end do
      end do
c
      return
      END

File: pkg/R/hankel.R
## ID: hankel.R, last updated 2025-05-26, F.Osorio

hankel <- function(x, y = NULL)
{ ## function returns the n-by-n Hankel matrix based on 'x' and 'y'
  if (!is.vector(x))
    stop("supply a vector-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric vector" )
  n <- length(x)

  if (is.null(y)) {
    y <- rep(0, n)
  } else {
    if (!is.vector(y))
      stop("supply a vector-like 'x'")
    if (!is.numeric(x))
      stop("argument x is not a numeric vector" )
    if (n != length(y))
      stop("arguments x and y do not have the same length.")
  }

  z <- .Fortran("hankel_mat",
                x = as.double(x),
                y = as.double(y),
                n = as.integer(n),
                mat = double(n * n),
                ldmat = as.integer(n),
                info = as.integer(0))
  if (z$info)
    stop(paste("hankel_mat gave error code", z$info))
  z <- matrix(z$mat, nrow = n, ncol = n)
  z
}
